Feature/HM-9-[Admin]-Create admin microservice

petarbelchev opened the pull request:
The Admin microservice has been created. It has two endpoints for now:
1. GET /Admins/brokers - Returns all users that have the Broker role;
2. POST /Admins/properties - Admins can attach .csv file with properties in format "Type;NumberOfRooms;District;Space;Floor;TotalFloorsInBuilding;SellerId;BrokerId(can be missing)" per row and per property.

On the Properties microservice, an "AddMultiple" endpoint has been created, which expects and inserts the properties passed from the Admins microservice.

petarbelchev committed:
The Admins microservice has been created.
Endpoint for retrieving all brokers has been created.

petarbelchev committed:
Merge pull request #13 from rosteslav/Develop

Merge Develop into feature/HM-9-Admin-Create-admin-microservice

petarbelchev committed:
The endpoint for adding multiple properties has been implemented.

petarbelchev committed:
An endpoint in the Admin microservice for adding multiple properties from .csv file has been created.
The endpoint for adding multiple properties in the Properties microservice has been changed to expect a collection of properties.

petarbelchev committed:
The request body for the AddMultiple endpoint in the Properties microservice has been set as required.

petarbelchev committed:
Duplicated 'Commands' folder has been removed.

petarbelchev committed:
The JWT argument in the SendProperties method has been changed to a string type.

rosteslav commented on src/Properties/Properties.Application/Features/Properties/Commands/AddMultipleProperties/AddMultiplePropertiesCommandHandler.cs:
```diff
@@ -0,0 +1,28 @@
+    {
+        private readonly IPropertiesRepository _propertiesRepository = propertiesRepository;
+        private readonly IMapper _mapper = mapper;
+        private readonly ILogger<AddMultiplePropertiesCommandHandler> _logger = logger;
```
Not used

petarbelchev replied:
Done!

rosteslav commented on src/Properties/Properties.Application/Features/Properties/Commands/AddMultipleProperties/AddMultiplePropertiesCommandHandler.cs:
```diff
@@ -0,0 +1,28 @@
+            var properties = _mapper.Map<IEnumerable<Property>>(request.Properties);
+            int result = await _propertiesRepository.AddMultiple(properties);
+
+            return new Response { Status = "Success", Message = $"{result} properties have been successfully added" };
```
Don't make a response here

petarbelchev replied:
Done!

rosteslav commented on src/Properties/Properties.Application/Features/Properties/Commands/AddMultipleProperties/AddMultiplePropertiesCommand.cs:
```diff
@@ -0,0 +1,10 @@
+
+namespace BuildingMarket.Properties.Application.Features.Properties.Commands.AddMultipleProperties
+{
+    public class AddMultiplePropertiesCommand : IRequest<Response>
```
Don't return response here

petarbelchev replied:
Done!

rosteslav commented on src/Properties/Properties.Api/Controllers/PropertiesController.cs:
```diff
@@ -37,6 +39,19 @@ public async Task<IActionResult> Add([FromBody] PropertyModel model)
+        public async Task<IActionResult> AddMultiple([Required][FromBody] IEnumerable<PropertyModel> properties)
+        {
+            var command = new AddMultiplePropertiesCommand { Properties = properties };
+            return Ok(await _mediator.Send(command));
```
Should bahave like add single, return nocontent

petarbelchev replied:
Done!

rosteslav commented on src/Properties/Properties.Application/Models/Response.cs:
```diff
@@ -0,0 +1,8 @@
+﻿namespace BuildingMarket.Properties.Application.Models
+{
+    public class Response
```
Not needed

petarbelchev replied:
Done!

rosteslav commented on src/Properties/Properties.Application/Features/Properties/Commands/AddMultipleProperties/AddMultiplePropertiesCommandHandler.cs:
```diff
@@ -0,0 +1,28 @@
+        public async Task<Response> Handle(AddMultiplePropertiesCommand request, CancellationToken cancellationToken)
+        {
+            var properties = _mapper.Map<IEnumerable<Property>>(request.Properties);
+            int result = await _propertiesRepository.AddMultiple(properties);
```
No return here - AddMultiple should be void (Task)

petarbelchev replied:
Done!

rosteslav commented on src/Properties/Properties.Infrasructure/Repositories/PropertiesRepository.cs:
```diff
@@ -26,6 +26,26 @@ public async Task Add(Property item)
+                _logger.LogError(ex, $"Error while adding multiple properties");
+            }
+
+            return result;
```
Behave like Add single

petarbelchev replied:
Done!

rosteslav commented on src/Properties/Properties.Infrasructure/Repositories/PropertiesRepository.cs:
```diff
@@ -26,6 +26,26 @@ public async Task Add(Property item)
+            try
+            {
+                await _context.Properties.AddRangeAsync(properties);
+                result = await _context.SaveChangesAsync();
```
no result -> if something's wrong -> exception

petarbelchev replied:
Done!

rosteslav commented on src/Properties/Properties.Infrasructure/Repositories/PropertiesRepository.cs:
```diff
@@ -26,6 +26,26 @@ public async Task Add(Property item)
+        public async Task<int> AddMultiple(IEnumerable<Property> properties)
+        {
+            _logger.LogInformation($"DB add multiple properties");
+            int result = 0;
```
Not needed

petarbelchev replied:
Done!

rosteslav commented on src/Properties/Properties.Application/Contracts/IPropertiesRepository.cs:
```diff
@@ -11,5 +11,7 @@ public interface IPropertiesRepository
 
         Task Add(Property item);
+
+        Task<int> AddMultiple(IEnumerable<Property> properties);
```
Taks, like Add

petarbelchev replied:
Done!

rosteslav commented on src/Admins/Admins.Api/Controllers/AdminsController.cs:
```diff
@@ -0,0 +1,54 @@
+    public class AdminsController(IMediator mediator, ILogger<AdminsController> logger, IMapper mapper) : ControllerBase
+    {
+        private readonly IMediator _mediator = mediator;
+        private readonly ILogger<AdminsController> _logger = logger;
```
log on every endpoing hit, like on other controllers

petarbelchev replied:
Done!

rosteslav commented on src/Properties/Properties.Api/Controllers/PropertiesController.cs:
```diff
@@ -37,6 +39,19 @@ public async Task<IActionResult> Add([FromBody] PropertyModel model)
+        [ProducesResponseType(StatusCodes.Status401Unauthorized)]
+        [ProducesResponseType(StatusCodes.Status403Forbidden)]
+        public async Task<IActionResult> AddMultiple([Required][FromBody] IEnumerable<PropertyModel> properties)
+        {
```
Log that the endpoint is hit

petarbelchev replied:
Done!

rosteslav commented on src/Admins/Admins.Api/Controllers/AdminsController.cs:
```diff
@@ -0,0 +1,54 @@
+                File = csvFile
+            });
+
+            return Ok(response);
```
Return NoContent, if error -> throw

petarbelchev replied:
Done!

rosteslav commented on src/Admins/Admins.Api/Program.cs:
```diff
@@ -0,0 +1,38 @@
+var configuration = builder.Configuration;
+services.AddControllers();
+services.AddEndpointsApiExplorer();
+services.AddIdentity<IdentityUser, IdentityRole>(options =>
```
We don't need this since we are not registering users here

petarbelchev replied:
Done!

rosteslav commented on src/Admins/Admins.Api/Program.cs:
```diff
@@ -0,0 +1,38 @@
+    options.Password.RequireNonAlphanumeric = false;
+    options.Password.RequiredLength = 3;
+})
+    .AddEntityFrameworkStores<AdminsDbContext>()
```
We are not doing any of that here, we work with tokens only

petarbelchev replied:
Done!

rosteslav commented on src/Admins/Admins.Api/Properties/launchSettings.json:
```diff
@@ -0,0 +1,49 @@
+      "environmentVariables": {
+        "ASPNETCORE_ENVIRONMENT": "Dev",
+        "ConnectionStrings__PostgresConnectionString": "Host=isabelle.db.elephantsql.com;Port=5432;Database=vvizkylm;Username=vvizkylm;Password=7mXXhXO5jLwqLNrCFx1Vmzt97nLcpabv;ApplicationName=BuildingMarket;",
+        "PropertiesMicroserviceHost": "http://localhost:5220",
```
We write directly to the DB

rosteslav replied:
Not needed

petarbelchev replied:
Done!

rosteslav commented on src/Admins/Admins.Application/Contracts/IAdminRepository.cs:
```diff
@@ -0,0 +1,9 @@
+{
+    public interface IAdminRepository
+    {
+        Task<IEnumerable<IdentityUser>> GetAllBrokers();
```
Add AddMultipleProperties as it is admin functionality

petarbelchev replied:
Done!

rosteslav commented on src/Admins/Admins.Application/Features/Admins/Commands/AddMultipleProperties/AddMultiplePropertiesCommand.cs:
```diff
@@ -0,0 +1,12 @@
+{
+    public class AddMultiplePropertiesCommand : IRequest<Response>
+    {
+        public string JWT { get; set; }
```
Lets not use JWT this way, not needed here

petarbelchev replied:
Done!

rosteslav commented on src/Admins/Admins.Application/Features/Admins/Commands/AddMultipleProperties/AddMultiplePropertiesCommandHandler.cs:
```diff
@@ -0,0 +1,105 @@
+        : IRequestHandler<AddMultiplePropertiesCommand, Response>
+    {
+        private readonly IConfiguration _configuration = configuration;
+        private readonly ILogger<AddMultiplePropertiesCommandHandler> _logger = logger;
```
Log something

petarbelchev replied:
Done!

rosteslav commented on src/Admins/Admins.Application/Features/Admins/Commands/AddMultipleProperties/AddMultiplePropertiesCommandHandler.cs:
```diff
@@ -0,0 +1,105 @@
+        }
+
+        private async Task<Response> SendProperties(IEnumerable<PropertyModel> properties, string jwt)
+        {
```
Don't send them anywhere, write it to the DB straightaway

rosteslav replied:
This method sets user id to be aur, and here we don't want that - admin bulk inserts the properties - he is not the seller of them. This is a helper function if we migrate from another site

petarbelchev replied:
Done!

rosteslav commented on src/Admins/Admins.Application/Features/Admins/Commands/AddMultipleProperties/AddMultiplePropertiesCommandHandler.cs:
```diff
@@ -0,0 +1,105 @@
+                string line = await streamReader.ReadLineAsync();
+                while (!string.IsNullOrWhiteSpace(line))
+                {
+                    string[] data = line.Split(';', StringSplitOptions.RemoveEmptyEntries);
```
User comma as a separator, let keep it right, CSV - comma separated values

petarbelchev replied:
Done!

rosteslav commented on src/Admins/Admins.Application/Features/Admins/Commands/AddMultipleProperties/AddMultiplePropertiesCommandHandler.cs:
```diff
@@ -0,0 +1,105 @@
+                Space = decimal.Parse(data[3]),
+                Floor = int.Parse(data[4]),
+                TotalFloorsInBuilding = int.Parse(data[5]),
+                SellerId = data[6]
```
BrokerId Also

petarbelchev replied:
Done!

rosteslav commented on src/Admins/Admins.Application/Features/Admins/Commands/AddMultipleProperties/AddMultiplePropertiesCommandHandler.cs:
```diff
@@ -0,0 +1,105 @@
+                SellerId = data[6]
+            };
+
+            if (data.Length > 7)
```
Make constants to make changes easy - Set constants in the beginning of the file:
- public const int TypePosition = 0;
- public const int TotalNumberOfPossition = 7;
this way when we make a change on fiestld just change the consts, not the logic

petarbelchev replied:
Done!

rosteslav commented on src/Admins/Admins.Api/Controllers/AdminsController.cs:
```diff
@@ -0,0 +1,54 @@
+            Request.Headers.TryGetValue("Authorization", out StringValues jwt);
+            var response = await _mediator.Send(new AddMultiplePropertiesCommand
+            {
+                JWT = jwt,
```
No JWT here

petarbelchev replied:
Done!

rosteslav commented on src/Admins/Admins.Api/Controllers/AdminsController.cs:
```diff
@@ -0,0 +1,54 @@
+        [ProducesResponseType(StatusCodes.Status403Forbidden)]
+        public async Task<IActionResult> AddMultipleProperties([Required][ValidCsvFile] IFormFile csvFile)
+        {
+            Request.Headers.TryGetValue("Authorization", out StringValues jwt);
```
Remove this since JWT will not be used in this way

petarbelchev replied:
Done!

rosteslav commented on src/Admins/Admins.Application/Models/Response.cs:
```diff
@@ -0,0 +1,8 @@
+﻿namespace BuildingMarket.Admins.Application.Models
+{
+    public class Response
```
Remove this all together

petarbelchev replied:
Done!

rosteslav commented on src/Admins/Admins.Infrastructure/Persistence/Configuration/ConnectionsConfig.cs:
```diff
@@ -0,0 +1,7 @@
+﻿namespace BuildingMarket.Admins.Infrastructure.Persistence.Configuration
+{
+    public class ConnectionsConfig
```
Move this to common and use it on all the other places as it's the same for every project

petarbelchev replied:
Done!

rosteslav commented on src/Admins/Admins.Infrastructure/Persistence/AdminsDbContext.cs:
```diff
@@ -0,0 +1,22 @@
+            base.OnModelCreating(modelBuilder);
+
+            modelBuilder.Entity<IdentityUser>().ToTable("Users", "security");
+            modelBuilder.Entity<IdentityRole>().ToTable("Roles", "security");
```
Don't need all of those.. only the user to read from. Maybe Roles and UserRoles as well, if  _userManager.GetUsersInRoleAsync(UserRoles.Broker) needs it. research what is needed

petarbelchev replied:
Done!

rosteslav commented on src/Admins/Admins.Infrastructure/Persistence/AdminsDbContext.cs:
```diff
@@ -0,0 +1,22 @@
+    {
+        protected override void OnModelCreating(ModelBuilder modelBuilder)
+        {
+            base.OnModelCreating(modelBuilder);
```
You will need properties as well in order to write to

petarbelchev replied:
Done!

rosteslav commented on src/Admins/Admins.Infrastructure/Repositories/AdminRepository.cs:
```diff
@@ -0,0 +1,30 @@
+    public class AdminRepository(UserManager<IdentityUser> userManager, ILogger<AdminRepository> logger) : IAdminRepository
+    {
+        private readonly ILogger<AdminRepository> _logger = logger;
+        private readonly UserManager<IdentityUser> _userManager = userManager;
```
Have Write to Properties as well, we'll do it here

petarbelchev replied:
Done!

petarbelchev committed:
The Admin microservice now inserts properties into the DB instead of using the Properties microservice;
Identity service registration in the Admin microservice has been refactored;
The Properties microservice returns NoContent or throws an exception when inserting multiple properties.

rosteslav commented on src/Admins/Admins.Infrastructure/Persistence/AdminsDbContext.cs:
```diff
@@ -0,0 +1,60 @@
+
+namespace BuildingMarket.Admins.Infrastructure.Persistence
+{
+    public class AdminsDbContext(DbContextOptions<AdminsDbContext> options) : IdentityDbContext<IdentityUser>(options)
```
Try If it works without being IdentityDbContext<IdentityUser>(options), just a DbContext. We are not registering users here after all, nor logging them

petarbelchev replied:
I tried this, but it doesn't work like that.

rosteslav commented on src/Admins/Admins.Application/Features/Admins/Commands/AddMultipleProperties/AddMultiplePropertiesCommandHandler.cs:
```diff
@@ -0,0 +1,86 @@
+            }
+        }
+
+        private async Task<IEnumerable<Property>> MapPropertiesFromFile(IFormFile file)
```
This logic should not be here - logic in infrastructure

petarbelchev replied:
Done!

rosteslav commented on src/Admins/Admins.Application/Features/Admins/Commands/AddMultipleProperties/AddMultiplePropertiesCommandHandler.cs:
```diff
@@ -0,0 +1,86 @@
+            return properties;
+        }
+
+        private static Property MapPropertyModel(string[] data)
```
This is also login for infrastructure

petarbelchev replied:
Done!

petarbelchev committed:
The logic for reading the CSV file has been moved into the Admin Repository.
